python/hail/ir/table_ir.py

```diff
@@ -21,18 +21,18 @@ def __init__(self, left, right, join_type, join_key):
         self.join_type = join_type
         self.join_key = join_key
 
-    def __str__(self):
+    def render(self, r):
         return '(TableJoin {} {} {} {})'.format(
-            escape_id(self.join_type), self.join_key, self.left, self.right)
+            escape_id(self.join_type), self.join_key, r(self.left), r(self.right))
 
 
 class TableUnion(TableIR):
     def __init__(self, children):
         super().__init__()
         self.children = children
 
-    def __str__(self):
-        return '(TableUnion {})'.format(' '.join([str(x) for x in self.children]))
+    def render(self, r):
+        return '(TableUnion {})'.format(' '.join([r(x) for x in self.children]))
```

Comment:
I find this motif a little weird. I don't understand why we need a `Renderer.__call__` method -- I would think that we can represent java IRs explicitly in Python (`JavaValueIR` `JavaTableIR` `JavaMatrixTableIR`) which know how to render themselves.

A design where this can be 
```
return '(TableUnion {})'.format(' '.join(x.render(r) for x in self.children))
```

seems nicer.

Reply:
The point here is that some IR are expensive to parse (import, read, etc.) so we want to do it once.  My design is to have the e.g. ReadMatrix node have a handle to the java object obtained by parsing it, and print that out as a JavaIR instead of rendering it as text and re-running it.  That's what `stop_at_jir` is about: any IR node, whatever type, should print as Java IR if has a stashed Java reference.

If I write it `x.render(r)` then everyone node has to check `stop_at_jir` and print the Java object (or just call through to a render function that does it.)

Reply:
hmmmm.

It seems like it can have that property but still have my proposed design -- when constructing a read / import node, we explicitly ask for a JavaMatrixIR node from Python.

It seems weird to have java IR nodes still maintain the whole IR stack, when they really just need a JavaMatrixIR (table, value). Does this make sense?

Reply:
Yes, I see what you're saying, but in the long term I think it's better to have knowledge of the original node in Python rather than the less transparent Java IR.

While that's my primary motivation, in the current design, until we have relational types implemented in Python, each new table construction will have to render/parse the IR to get the types in Python MatrixTable/Table.  So either you have to turn every node into a Java node or constructing relational pipelines becomes quadratic.

If you still disagree, I will change it.